File: javascript/manipulate.js
/**
 *
 * @param {string} text string of value example "aabb", "abcde"
 * @returns {string[]} string array of result
 */

const factorial = (number) => {
  if (typeof number !== "number") throw new Error("Input must be number");

  let answer = 1;

  if (number !== 0) {
    for (let i = 2; i <= number; i++) {
      answer = answer * i;
    }
  }

  return answer;
};

const manipulate = (text) => {
  if (typeof text !== "string") {
    return "Input must be string";
  }

  const result = [];

  const splitChar = text.split("");
  for (let index = 0; index < splitChar.length; index++) {
    let char = splitChar[index];
    const otherChar = splitChar.filter((_, i) => index !== i);

    for (let loop = 0; loop < factorial(otherChar.length); loop++) {
      if (loop) {
        const firstElement = otherChar.shift();
        otherChar.push(firstElement);
      }
      const combine = char + otherChar.join("");
      if (!result.includes(combine)) result.push(combine);
    }
  }

  return result;
};

module.exports = {
  manipulate,
};

File: golang/manipulate.go
package main

import (
	"strings"
)

func factorial(number int) int {
	answer := 1
	if number >= 2 {
		for i := 2; i <= number; i++ {
			answer = answer * i
		}
	}
	return answer
}

func manipulate(text string) []string {
	result := make([]string, 0)
	splitChar := strings.Split(text, "")

	for index := 0; index < len(splitChar); index++ {
		char := splitChar[index]
		otherChar := make([]string, 0)
		for i, ch := range splitChar {
			if i != index {
				otherChar = append(otherChar, ch)
			}
		}

		for loop := 0; loop < factorial(len(otherChar)); loop++ {
			if loop > 0 {
				firstElement := otherChar[0]
				otherChar = append(otherChar[1:], firstElement)
			}
			combine := char + strings.Join(otherChar, "")
			if !contains(result, combine) {
				result = append(result, combine)
			}
		}
	}

	return result
}

func contains(arr []string, str string) bool {
	for _, a := range arr {
		if a == str {
			return true
		}
	}
	return false
}
